File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    let count = 0;
    const visited = new Set();
    
    function dfs(node) {
        visited.add(node);
        for (let i = 0; i < computers[node].length; i++) {
            if (node === i) continue;
            if (!visited.has(i) && computers[node][i] === 1) dfs(i);
        }
    }
    
    for (let i = 0; i < n; i++) {
        if (visited.has(i)) continue;
        count++;
        dfs(i);
    }
    
    return count;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
# def solution(n, computers):
#     cnt = 0
#     visited = [False] * n
    
#     def dfs(node):
#         visited[node] = True
        
#         for i, nxt in enumerate(computers[node]):
#             if nxt == 1 and not visited[i]:
#                 dfs(i)
    
#     for i in range(n):
#         if not visited[i]:
#             cnt += 1
#             dfs(i)
            
#     return cnt


def solution(n, computers):
    visited = [False] * n
    cnt = 0
    
    def dfs(node):
        visited[node] = True
        
        for i in range(n):
            if node == i: continue
            if computers[node][i] == 1:
                if not visited[i]:
                    dfs(i)
                
    for i in range(n):
        if not visited[i]:
            dfs(i)
            cnt += 1
    
    return cnt